src/sensor_fusion.c

```diff
@@ -12,73 +13,39 @@
 #include <gsl/gsl_eigen.h>
 #include "sensor_fusion.h"
 
-/**
- * \brief Structure for storing support degree matrix
- * \details For the given number of sensors the structure will store support degree matrix
- */
-struct support_degree_matrix {
-    int sensor_count;
-    double *sd_matrix;
-};
-
-/**
- * \brief Structure for storing eigen value and eigen vector
- * \details For a given support degree matrix of
- * sensors the corresponding eigen values and vectors are stored in structure
- */
-struct eigen_value_vector {
-    double *eigen_value;
-    double **eigen_vector;
-};
-
-/**
- * \brief Structure to store computed support degree matrix
- * \details Structure pointer for calculating support degree matrix
- * \param[in] sensor_readings array of sensor readings
- * \return Success: pointer to the structure support_degree_matrix
- * \return Failure: NULL
- */
-struct support_degree_matrix *compute_support_degree_matrix(float *sensor_readings);
-
-/**
- * \brief Structure to store computed eigen value and eigen vector
- * \details Structure pointer for calculating eigen value and eigen vector
- * \param[in] support_degree_matrix pointer to structure support degree matrix
- * \return Success: pointer to structure eigen_value_vector
- * \return Failure: NULL
- */
-struct eigen_value_vector *compute_eigen(struct support_degree_matrix *spd);
-
 /**
  * \brief Structure pointer for computed support degree matrix
  * \details Structure pointer for calculating support degree matrix
  * \param[in] sensor_readings array of sensor readings
  * \return Success: pointer to the structure support_degree_matrix
  * \return Failure: NULL
  */
-struct support_degree_matrix *compute_support_degree_matrix(float *sensor_readings) {
+struct support_degree_matrix *compute_support_degree_matrix(double *sensor_readings) {
     struct support_degree_matrix *spd;
     spd = (struct support_degree_matrix *)malloc(sizeof(struct support_degree_matrix));
+    // count the number of sensor readings
     int length = sizeof(sensor_readings) / sizeof(sensor_readings[0]);
```

Comment:
I'm fairly certain this is a logical error in your code. sizeof(sensor_readings) will return the size of a pointer and sizeof(sensor_readings[0]) will return the size of a double. Instead you should pass the size of the array as a function parameter.

Reply:
Yes, the array length can be calculated outside the sensor_fusion.c file and passed as a parameter. However, from the software engineering point of view we wish to have a good code maintenance for future work. Therefore, we kept the logic to calculate sensor length within the sensor file. So that all the relevant logic is self-contained in one place and not scattered across different files. Also, the self-contained code helps in code portability, scalability and robustness. 

The calculation is correct because parameter to the function is => **_(double *sensor_readings)_** and the command => **_sizeof(sensor_readings)_** will return size of the array of type double. It will not return size of pointer. By dividing it with first element of array, we get length of array (aka length of sensor readings.) 

The length of sensor readings coming from CSV file is dynamic for given timestamp. There can be 5 sensor readings for time T1 and there can be 3 sensor readings for time T2. Therefore, we calculate dynamic sensor reading length inside the sensor function and not in main.c file.

Thanks!